File: src/main/java/io/github/kingg22/api/vacunas/panama/web/controller/BulkController.java
package io.github.kingg22.api.vacunas.panama.web.controller;

import io.github.kingg22.api.vacunas.panama.response.ApiResponse;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory;
import io.github.kingg22.api.vacunas.panama.response.ApiResponseUtil;
import io.github.kingg22.api.vacunas.panama.service.IPacienteService;
import io.github.kingg22.api.vacunas.panama.service.IUsuarioManagementService;
import io.github.kingg22.api.vacunas.panama.web.dto.PacienteDto;
import io.github.kingg22.api.vacunas.panama.web.dto.RegisterUser;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.jetbrains.annotations.NotNull;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.context.request.ServletWebRequest;

@Slf4j
@RestController
@RequiredArgsConstructor
@RequestMapping(path = "/vacunacion/v1/bulk", produces = MediaType.APPLICATION_JSON_VALUE)
public class BulkController {
    private final IUsuarioManagementService usuarioManagementService;
    private final IPacienteService pacienteService;

    @Transactional
    @PostMapping("/paciente-usuario-direccion")
    public ResponseEntity<ApiResponse> createPacienteUsuario(
            @NotNull @RequestBody @Valid PacienteDto pacienteDto, ServletWebRequest request) {
        var apiResponse = ApiResponseFactory.createResponse();
        log.debug(pacienteDto.toString());
        var validateContent = this.pacienteService.validateCreatePacienteUsuario(pacienteDto);
        apiResponse.addErrors(validateContent.getErrors());
        apiResponse.addWarnings(validateContent.getWarnings());
        if (apiResponse.hasErrors()) {
            apiResponse.addStatusCode(HttpStatus.BAD_REQUEST);
            return ApiResponseUtil.sendResponse(apiResponse, request);
        }
        var pacienteContent = this.pacienteService.createPaciente(pacienteDto);
        apiResponse.addWarnings(pacienteContent.getWarnings());
        apiResponse.addErrors(pacienteContent.getErrors());
        if (pacienteContent.hasErrors()) {
            apiResponse.addStatusCode(HttpStatus.BAD_REQUEST);
            return ApiResponseUtil.sendResponse(apiResponse, request);
        }
        var registerUser = new RegisterUser(
                pacienteDto.persona().usuario(),
                pacienteDto.persona().cedula(),
                pacienteDto.persona().pasaporte());
        var apiContentResponse = this.usuarioManagementService.createUser(registerUser);
        apiResponse.mergeContentResponse(apiContentResponse);
        if (apiContentResponse.hasErrors()) {
            apiResponse.addStatusCode(HttpStatus.BAD_REQUEST);
        } else {
            apiResponse.addStatusCode(HttpStatus.CREATED);
        }
        return ApiResponseUtil.sendResponse(apiResponse, request);
    }
}

File: src/main/kotlin/io/github/kingg22/api/vacunas/panama/controller/BulkController.kt
package io.github.kingg22.api.vacunas.panama.controller

import io.github.kingg22.api.vacunas.panama.modules.paciente.dto.PacienteInputDto
import io.github.kingg22.api.vacunas.panama.modules.paciente.service.PacienteService
import io.github.kingg22.api.vacunas.panama.modules.usuario.dto.RegisterUserDto
import io.github.kingg22.api.vacunas.panama.modules.usuario.service.UsuarioService
import io.github.kingg22.api.vacunas.panama.response.ApiResponseFactory.createResponse
import io.github.kingg22.api.vacunas.panama.response.ApiResponseUtil.createResponseEntity
import io.github.kingg22.api.vacunas.panama.util.logger
import io.vertx.ext.web.RoutingContext
import jakarta.validation.Valid
import jakarta.ws.rs.POST
import jakarta.ws.rs.Path
import jakarta.ws.rs.Produces
import jakarta.ws.rs.core.MediaType
import jakarta.ws.rs.core.Response

/** TODO to split in modules controllers */
@Path("/bulk")
@Produces(MediaType.APPLICATION_JSON)
class BulkController(private val usuarioService: UsuarioService, private val pacienteService: PacienteService) {
    private val log = logger()

    @POST
    @Path("/paciente-usuario-direccion")
    suspend fun createPacienteUsuario(@Valid pacienteInputDto: PacienteInputDto, rc: RoutingContext): Response {
        val apiResponse = createResponse()
        log.debug("Received a request to create a new Paciente, Dirección and User: {}", pacienteInputDto.toString())
        val pacienteDto = pacienteInputDto.toPacienteDto()
        log.debug(pacienteDto.toString())
        val pacienteContent = pacienteService.createPaciente(pacienteDto)
        log.trace(pacienteContent.toString())
        apiResponse.addWarnings(pacienteContent.warnings)
        apiResponse.addErrors(pacienteContent.errors)
        if (pacienteContent.hasErrors()) {
            log.trace("CreatePaciente return errors: {}", pacienteContent.errors)
            apiResponse.addStatusCode(400)
            return createResponseEntity(apiResponse, rc)
        }
        val registerUserDto = RegisterUserDto(
            pacienteDto.persona.usuario!!,
            pacienteDto.persona.cedula,
            pacienteDto.persona.pasaporte,
        )
        log.trace("RegisterUserDto: {}", registerUserDto.toString())
        // TODO add authentication verify of scopes
        val apiContentResponse = usuarioService.createUser(registerUserDto)
        apiResponse.mergeContentResponse(apiContentResponse)
        log.trace("CreateUser return: {}", apiContentResponse.toString())
        if (apiContentResponse.hasErrors()) {
            apiResponse.addStatusCode(400)
        } else {
            apiResponse.addStatusCode(201)
        }
        return createResponseEntity(apiResponse, rc)
    }
}
